ext/gvl_tracing_native_extension/gvl_tracing.c

```diff
@@ -49,21 +49,62 @@
   #define UNUSED_ARG
 #endif
 
-static VALUE tracing_start(VALUE _self, VALUE output_path);
-static VALUE tracing_stop(VALUE _self);
-static double timestamp_microseconds(void);
-static void set_native_thread_id(void);
-static void render_event(const char *event_name);
-static void on_thread_event(rb_event_flag_t event, const rb_internal_thread_event_data_t *_unused1, void *_unused2);
-static void on_gc_event(VALUE tpval, void *_unused1);
-static VALUE mark_sleeping(VALUE _self);
+typedef struct {
+  bool current_thread_seen;
+  unsigned int current_thread_serial;
+  uint64_t thread_id;
+  VALUE thread;
+  rb_event_flag_t previous_state; // Used to coalesce similar events
+  bool sleeping; // Used to track when a thread is sleeping
+} thread_local_state;
+
+#ifdef HAVE_RB_INTERNAL_THREAD_SPECIFIC_GET // 3.3+
+
+static int thread_storage_key = 0;
+
+static size_t thread_local_state_memsize(UNUSED_ARG const void *data) {
+    return sizeof(thread_local_state);
+}
+
+static void thread_local_state_mark(void *data) {
+  thread_local_state *state = (thread_local_state *)data;
+  rb_gc_mark(state->thread);
+}
+
+static const rb_data_type_t thread_local_state_type = {
+    .wrap_struct_name = "GvlTracing::__threadLocal",
+    .function = {
+        .dmark = thread_local_state_mark,
+        .dfree = RUBY_DEFAULT_FREE,
+        .dsize = thread_local_state_memsize,
+    },
+    .flags = RUBY_TYPED_FREE_IMMEDIATELY | RUBY_TYPED_WB_PROTECTED,
```

Comment:
So my VM-foo fails me slightly on this part... `RUBY_TYPED_WB_PROTECTED` is ok here since we never perform any reads or writes to Ruby objects, is that correct?